### PR TITLE
[NG] Datagrid chocolate

### LINKED ISSUE #623: clr-datagrid action-overflow not visible and formatting issues with dynamic data
**Select one ...**  (check one with "x")
```
[x] bug
[ ] feature request
[ ] enhancement
```

### Expected behavior
The datagrid should work flawlessly even for dynamic data.

### Actual behavior
If i have data coming in from server, the datagrid does not show the action-overflow column for the grid and the columns are also highly misaligned.
Only when some user-action triggers something in angular does the datagrid re-render properly as expected.
1. In the plunker, wait for two seconds for the first grid to load data.
2. See the problems. 
3. Then just hover over the green button which has a bound (mousehover) to trigger angular.
4. Watch how the first grid re-renders correctly

### Reproduction of behavior

https://plnkr.co/edit/J5oMyeVWgGWJFTrg9B73

### Environment details

* **Angular version:** 2.0.X

* **Clarity version:** 0.8.11

* **OS and version:** 

* **Browser:** all

### PR DESCRIPTION
**Pull request by youdz**
Cleaning up some lifecycle, adding a generic solution to projected
siblings updating each other or template children updating the parent
view when created.

We will have to be very careful, any future commit using or modifying
any file in the chocolate factory should be properly monitored
and evaluated. The valid use cases for it are rare, it turns out the
hideable columns didn't even need it, it was just a case of messed
up lifecycle where everything happened after the view was already
initialized. Beware! The chocolate factory is much scarier than the
name indicates. Well, not as scary as Johnny Depp's Willy Wonka, but
still scary.

Fixes #623, #744, #817, #862, #928, #950 and #951 (not counting duplicates).

### CODE REVIEW

File: src/app/app.html [hunk 1]
```diff
@@ -22,17 +22,17 @@
...
-                    <clr-icon shape="cog"></clr-icon>
-                    <clr-icon shape="caret down"></clr-icon>
-                </button>
-                <clr-dropdown-menu>
-                    <a href="javascript://" clrDropdownItem>About</a>
-                    <a href="javascript://" clrDropdownItem>Preferences</a>
-                    <a href="javascript://" clrDropdownItem>Log out</a>
-                </clr-dropdown-menu>
-            </clr-dropdown>
+            <!--<clr-dropdown [clrMenuPosition]="'bottom-right'">-->
```
Comment (mathisscott) [inline src/app/app.html]: Any reason we want to keep this in here? Can't we just remove the dropdown?

File: src/app/datagrid/basic-structure/basic-structure.html [hunk 2]
```diff
@@ -15,20 +15,43 @@
...
+            <clr-dg-row-detail *clrIfExpanded>Detail</clr-dg-row-detail>
+        </ng-template>
     </clr-dg-row>
 
-    <clr-dg-footer>{{users.length}} users</clr-dg-footer>
+    <clr-dg-footer>{{items.length}} items</clr-dg-footer>
 </clr-datagrid>
 
-<clr-example [clrCode]="example" clrLanguage="html"></clr-example>
+<!--<clr-datagrid>-->
```
Comment (mathisscott) [inline src/app/datagrid/basic-structure/basic-structure.html]: +1

File: src/clarity-angular/data/datagrid/_datagrid.clarity.scss [hunk 3]
```diff
@@ -151,9 +151,6 @@
 
         .datagrid-column, .datagrid-cell {
             text-align: left;
-        }
-
-        clr-dg-column {
```
Comment (mathisscott) [inline src/clarity-angular/data/datagrid/_datagrid.clarity.scss]: 👍

File: src/clarity-angular/data/datagrid/chocolate/datagrid-willy-wonka.ts [hunk 4]
```diff
@@ -0,0 +1,8 @@
+import {Directive} from "@angular/core";
+import {WillyWonka} from "../../../utils/chocolate/willy-wonka";
+
+@Directive({
+    selector: "clr-datagrid",
+    providers: [{provide: WillyWonka, useExisting: DatagridWillyWonka}]
+})
+export class DatagridWillyWonka extends WillyWonka {}
```
Comment (mathisscott) [inline src/clarity-angular/data/datagrid/chocolate/datagrid-willy-wonka.ts]: :1st_place_medal:

File: src/clarity-angular/data/datagrid/datagrid-column-toggle.spec.ts [hunk 5]
```diff
@@ -84,13 +84,13 @@ export default function (): void {
...
+                let testEvent = false;
 
-                expect(testColumn.hidden).toBe(true);    // hidden
+                expect(testColumn.hidden).toBe(false);    // showing
                 component.toggleColumn(testEvent, testColumn);
-                expect(testColumn.hidden).toBe(false);   // showing
+                expect(testColumn.hidden).toBe(true);   // hidden
                 component.toggleColumn(!testEvent, testColumn);
-                expect(testColumn.hidden).toBe(true);    // hidden
+                expect(testColumn.hidden).toBe(false);    // showing
```
Comment (mathisscott) [inline src/clarity-angular/data/datagrid/datagrid-column-toggle.spec.ts]: 👍

File: src/clarity-angular/data/datagrid/datagrid-detail-registerer.ts [hunk 6]
```diff
@@ -0,0 +1,23 @@
+import {Directive, Optional} from "@angular/core";
+import {ExpandableRowsCount} from "./providers/global-expandable-rows";
+
+/*
+ * I don't think this deserves to be in IfExpanded itself,
```
Comment (mathisscott) [inline src/clarity-angular/data/datagrid/datagrid-detail-registerer.ts]: 💯

File: src/clarity-angular/data/datagrid/datagrid-row.spec.ts [hunk 7]
```diff
@@ -193,16 +192,15 @@ export default function(): void {
                 expect(context.clarityElement.textContent).not.toMatch("Detail");
             });
 
-            // FIXME: PhantomJS being a d***
-            xit("displays both the row and the details when expanded and not replacing", fakeAsync(function () {
```
Comment (mathisscott) [inline src/clarity-angular/data/datagrid/datagrid-row.spec.ts]: ❤️

File: src/clarity-angular/data/datagrid/datagrid.spec.ts [hunk 8]
```diff
@@ -264,24 +255,48 @@ export default function(): void {
...
+            });
+
+            describe("ngFor", function() {
+                it("doesn't taunt with chocolate on actionable rows", function() {
+                    let context = this.create(Datagrid, ChocolateNgForTest);
+                    context.testComponent.action = true;
+                    expect(() => context.detectChanges()).not.toThrow();
+                });
+
+                it("doesn't taunt with chocolate on expandable rows", function() {
```
Comment (mathisscott) [inline src/clarity-angular/data/datagrid/datagrid.spec.ts]: :1st_place_medal:

File: src/app/app.html (hunk 1 above)
Comment (youdz) [inline src/app/app.html]: My bad, this is clean-up that I missed on my second commit. Thanks for pointing it out, I'm rolling back to the current state in master.

**Pull request merged**